File: job_copilot/app/integrations/adzuna.py
"""Adzuna job search integration."""
from __future__ import annotations

from datetime import datetime
from typing import Any

import httpx

from app.core.config import get_settings
from app.core.logging import get_logger
from app.models.schemas import JobPosting

log = get_logger(__name__)

BASE_URL = "https://api.adzuna.com/v1/api/jobs/us/search"


async def search_jobs(
    query: str = "Data Engineer",
    location: str = "New Jersey",
    posted_within_days: int = 7,
    page: int = 1,
) -> list[JobPosting]:
    settings = get_settings()
    if not settings.adzuna_app_id or not settings.adzuna_api_key:
        log.warning("adzuna_skipped", reason="ADZUNA credentials not set")
        return []

    params = {
        "app_id": settings.adzuna_app_id,
        "app_key": settings.adzuna_api_key,
        "results_per_page": 50,
        "what": query,
        "where": location,
        "max_days_old": posted_within_days,
        "sort_by": "date",
        "content-type": "application/json",
    }

    jobs: list[JobPosting] = []
    async with httpx.AsyncClient(timeout=30) as client:
        try:
            resp = await client.get(f"{BASE_URL}/{page}", params=params)
            resp.raise_for_status()
            data = resp.json()
            for item in data.get("results", []):
                jobs.append(_normalize(item))
            log.info("adzuna_results", count=len(jobs))
        except Exception as e:
            log.error("adzuna_error", error=str(e))
    return jobs


def _normalize(item: dict[str, Any]) -> JobPosting:
    posted_at = None
    if item.get("created"):
        try:
            posted_at = datetime.fromisoformat(item["created"].replace("Z", "+00:00"))
        except (ValueError, TypeError):
            pass

    location_parts = []
    loc = item.get("location", {})
    for area in loc.get("area", []):
        location_parts.append(area)
    location_str = ", ".join(location_parts) if location_parts else ""

    salary = ""
    if item.get("salary_min") and item.get("salary_max"):
        salary = f"${item['salary_min']:,.0f} - ${item['salary_max']:,.0f}"

    return JobPosting(
        id=f"adzuna_{item.get('id', '')}",
        title=item.get("title", ""),
        company=item.get("company", {}).get("display_name", ""),
        location=location_str,
        remote="remote" in item.get("title", "").lower() or "remote" in item.get("description", "").lower(),
        posted_at=posted_at,
        source="adzuna",
        url=item.get("redirect_url", ""),
        jd_text=item.get("description", ""),
        salary=salary,
        company_domain="",
        raw_data=item,
    )

File: web/src/lib/job-sources/adzuna.ts
import type { JobPosting } from "../types";

export async function searchAdzuna(
  query = "Data Engineer",
  location = "New Jersey",
  postedWithinDays = 7
): Promise<JobPosting[]> {
  const appId = process.env.ADZUNA_APP_ID;
  const apiKey = process.env.ADZUNA_API_KEY;
  if (!appId || !apiKey) return [];

  const params = new URLSearchParams({
    app_id: appId,
    app_key: apiKey,
    results_per_page: "50",
    what: query,
    where: location,
    max_days_old: String(postedWithinDays),
    sort_by: "date",
    "content-type": "application/json",
  });

  try {
    const resp = await fetch(`https://api.adzuna.com/v1/api/jobs/us/search/1?${params}`);
    if (!resp.ok) return [];
    const data = await resp.json();
    return (data.results || []).map(normalize);
  } catch {
    return [];
  }
}

function normalize(item: Record<string, unknown>): JobPosting {
  const loc = item.location as Record<string, unknown> | undefined;
  const areas = (loc?.area as string[]) || [];
  const locationStr = areas.join(", ");
  const title = (item.title as string) || "";
  const desc = (item.description as string) || "";

  let salary = "";
  const minS = item.salary_min as number | undefined;
  const maxS = item.salary_max as number | undefined;
  if (minS && maxS) salary = `$${minS.toLocaleString()} - $${maxS.toLocaleString()}`;

  return {
    id: `adzuna_${item.id || ""}`,
    title,
    company: ((item.company as Record<string, string>)?.display_name) || "",
    location: locationStr,
    remote: title.toLowerCase().includes("remote") || desc.toLowerCase().includes("remote"),
    posted_at: (item.created as string) || null,
    source: "adzuna",
    url: (item.redirect_url as string) || "",
    jd_text: desc,
    salary,
    visa_signal: "unknown",
    company_domain: "",
    raw_data: item,
  };
}
